File: Leetcode - Top Interview Questions/238. Product of Array Except Self.cpp
class Solution
{
public:
    int mul(vector<int> nums, int idx)
    {
        int ans = 1;
        for (int i = 0; i < nums.size(); i++)
        {
            if (i != idx)
                ans *= nums[i];
        }
        return ans;
    }
    vector<int> productExceptSelf(vector<int> &nums)
    {
        vector<int> res;
        int ans = mul(nums, -1);
        for (int i = 0; i < nums.size(); i++)
        {
            if (nums[i] == 0)
                res.push_back(mul(nums, i));
            else
                res.push_back(ans / nums[i]);
        }
        return res;
    }
};

File: Leetcode - Top Interview Questions/238.Product_of_array_except_self.java
class Solution 
{
    public int[] productExceptSelf(int[] nums) 
    {
        int[] rmul = Arrays.copyOf(nums, nums.length);
        int[] lmul = Arrays.copyOf(nums, nums.length);
        
        for(int i=1,j = nums.length-2; i< nums.length; i++, j--)
        {
            
            rmul[i] *= rmul[i-1];
            lmul[j] *= lmul[j+1];
        }
        
        for(int i=0;i<nums.length;i++)
        {
            if(i == 0)
            {
                nums[i] = lmul[i+1];
            }
            else if(i == nums.length-1)
            {
                nums[i] = rmul[i-1];
            }
            else
            {
                nums[i] = rmul[i-1] * lmul[i+1];
            }
        }
        
        return nums;
    }
}
